api/pkg/transformer/feast/feature_retriever.go

```diff
@@ -213,7 +214,29 @@ func (fr *FeastRetriever) buildEntityRows(ctx context.Context, symbolRegistry sy
 		}
 	}
 
-	return entities, nil
+	uniqueEntities, err := dedupEntities(entities)
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueEntities, nil
+}
+
+func dedupEntities(rows []feast.Row) ([]feast.Row, error) {
+	uniqueRows := make([]feast.Row, 0, len(rows))
+	rowLookup := make(map[uint64]bool)
+	for _, row := range rows {
+		rowByte, err := json.Marshal(row)
```

Comment:
`json.Marshal` might be too expensive for this.
Alternatively we can "stringify" the row. I.e. create a string by appending all the keys and values from the Row.
The string then can be used as key of `rowLookup`
```
rowLookup := make(map[string]bool)
rowLookup[stringify(row)] = true
```

Reply:
there is no guarantee that keys in maps is sorted, so we need to sort the keys first. Let me check the performance of it

Reply:
I add this 
`func dedupEntities_stringify(rows []feast.Row) ([]feast.Row, error) {
	uniqueRows := make([]feast.Row, 0, len(rows))
	rowLookup := make(map[string]bool)
	keys := make([]string, 0)
	if len(rows) == 0 {
		return rows, nil
	}
	for key := range rows[0] {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, row := range rows {
		var keyString strings.Builder
		for _, k := range keys {
			keyString.WriteString(fmt.Sprintf("%s:%v_", k, row[k]))
		}
		rowKey := keyString.String()
		if _, found := rowLookup[rowKey]; !found {
			uniqueRows = append(uniqueRows, row)
			rowLookup[rowKey] = true
		}
	}
	return uniqueRows, nil
}`

`func Benchmark_dedup_stringify(b *testing.B) {
	b.StopTimer()

	entities := make([]feast.Row, 0, 100)
	for i := 0; i < 100; i++ {
		row := feast.Row{}
		row["merchant_id"] = feast.Int32Val(int32(i))
		// row["driver_id"] = feast.Int32Val(int32((i + 1) * 2))
		entities = append(entities, row)
	}
	b.StartTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		request, _ := dedupEntities_stringify(entities)
		_ = request
	}
}`

benchmark result:
`goos: darwin
goarch: amd64
pkg: github.com/gojek/merlin/pkg/transformer/feast
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
Benchmark_dedup_stringify-8   	    8710	    135942 ns/op	   31236 B/op	     812 allocs/op`

Reply:
also compare with existing dedup method

```go
func Benchmark_dedup(b *testing.B) {
	b.StopTimer()

	entities := make([]feast.Row, 0, 100)
	for i := 0; i < 100; i++ {
		row := feast.Row{}
		row["merchant_id"] = feast.Int32Val(int32(i))
		// row["driver_id"] = feast.Int32Val(int32((i + 1) * 2))
		entities = append(entities, row)
	}
	b.StartTimer()
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		request, _ := dedupEntities(entities)
		_ = request
	}
}
```

goos: darwin
goarch: amd64
pkg: github.com/gojek/merlin/pkg/transformer/feast
cpu: Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
Benchmark_dedup-8   	   13713	     87502 ns/op	   29856 B/op	     618 allocs/op

Reply:
Got it, I am wondering if there is a way to convert `feast.Row` into `byte[]` which can be more efficient than using json.